File: cafw-backend/app/firewall/logger.py
from sqlalchemy.orm import Session
from app.models import AttackLog, BlockedIP

def log_attack(
        db: Session,
        ip_address: str,
        method: str,
        endpoint: str,
        category: str,
        payload: str,
        rule_matched: str,
        action: str = "blocked"
):
    """Saves a blocked attack event to the database."""
    log = AttackLog(
        ip_address=ip_address,
        method=method,
        endpoint=endpoint,
        category=category,
        payload=payload,
        rule_matched=rule_matched,
        action=action
    )
    db.add(log)
    db.commit()
    db.refresh(log)
    return log


def is_ip_blocked(db: Session, ip_address: str) -> bool:
    """Checks if an IP is in the permanent block list."""
    blocked = db.query(BlockedIP).filter(
        BlockedIP.ip_address == ip_address,
        BlockedIP.is_active == True
    ).first()
    return blocked is not None


def block_ip(db: Session, ip_address: str, reason: str):
    """Permanently blocks an IP address."""
    existing = db.query(BlockedIP).filter(
        BlockedIP.ip_address == ip_address
    ).first()
    if existing:
        existing.is_active = True
        existing.reason = reason
    else:
        blocked = BlockedIP(
            ip_address=ip_address,
            reason=reason
        )
        db.add(blocked)
    db.commit()

File: cafw-backend/firewall/logger.js
const { execute } = require('../database');

/**
 * Log an attack to the database.
 * @param {object} params
 * @param {string} params.ip - Client IP
 * @param {string} params.method - HTTP method
 * @param {string} params.endpoint - Requested endpoint
 * @param {string} params.category - Attack category (SQL Injection, XSS, etc.)
 * @param {string} params.payload - The malicious payload
 * @param {string} params.ruleMatched - The regex pattern that matched
 */
function logAttack({ ip, method, endpoint, category, payload, ruleMatched }) {
  try {
    execute(
      `INSERT INTO attack_logs (ip_address, method, endpoint, category, payload, rule_matched, action)
       VALUES (?, ?, ?, ?, ?, ?, 'blocked')`,
      [ip, method, endpoint, category || 'Unknown', payload || '', ruleMatched || '']
    );
  } catch (err) {
    console.error('Failed to log attack:', err.message);
  }
}

/**
 * Log a request (blocked or allowed) to request_logs.
 */
function logRequest({ ip, method, endpoint, blocked, category }) {
  try {
    execute(
      `INSERT INTO request_logs (ip_address, method, endpoint, blocked, category)
       VALUES (?, ?, ?, ?, ?)`,
      [ip, method, endpoint, blocked ? 1 : 0, category || null]
    );
  } catch (err) {
    console.error('Failed to log request:', err.message);
  }
}

module.exports = { logAttack, logRequest };
